### ISSUE TITLE
Model.save() method should not always populate associations

### ISSUE THREAD

**Issue by mphasize** [balderdashy/waterline#441]
The model.save() instance method should accept a parameter to indicate whether the model associations should be populated after saving.

Following use case: Implementing a publish method on a post model

In PostController I added a `publish` action loosely based off the `update` blueprint.

	publish: function ( req, res ) {
		var pk = req.options.id || ( req.options.where && req.options.where.id ) || req.param( 'id' );

		var values = null;
                // the json-structure used here is closer to json-api specs with { "post" : { ... values ... }}
		if ( req.body && req.body[ "post" ] ) {
			values = req.body[ "post" ];
			// Omit any params w/ undefined values
			values = _.omit( values, function ( p ) {
				if ( _.isUndefined( p ) ) return true;
			} );
		}
		if ( values === null ) {
			return res.serverError( "Please provide values for 'post' to be updated in the JSON API format, e.g. { post: { ... values ... } }" );
		}

		Post.findOne( pk ).exec( function found( err, post ) {
				if ( err ) return res.serverError( err );
				if ( !post ) return res.notFound();

				post.publish( values, function ( err, publishedPost ) {
					if ( err ) return res.negotiate( err );

					// If we have the pubsub hook, use the Model's publish method
					// to notify all subscribers about the update.
					if ( req._sails.hooks.pubsub ) {
						if ( req.isSocket ) {
							Post.subscribe( req, [ publishedPost ] );
						}
						Post.publishUpdate( pk, _.cloneDeep( publishedPost.toJSON() ), !req.options.mirror && req, {
							previous: post.toJSON()
						} );
					}

					var jsonAPI = {};
					jsonAPI[ "post" ] = publishedPost;
					res.ok( jsonAPI );

				} );
			}

		);

After **Post.findOne()** the returned `post` has flat (i.e. not populated) associations with only respective IDs returned.

After **post.save()** (called in post.publish()) the associations are populated in `publishedPost`.

This leads to an error in **Post.publishUpdate** with `previous` set, because the model state saved in previous doesn't have populated associations and the current state has.

The error is generated in https://github.com/balderdashy/sails/blob/master/lib/hooks/pubsub/index.js#L977

    if ( invalidId || invalidAlias || invalidAddedId ) {
      return sails.log.error(
        'Invalid usage of ' + this.identity +
        '`publishAdd(id, alias, idAdded, [socketToOmit])`'
    );

A work-around would be to populate the findOne query or to delete populated associations after the .save(), but both seem inefficient.

That's why I want to propose to add an option to the `.save` method, that modifies the behaviour of .save() in https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L168

    // Build up a new query and re-populate everything
    var query = context.findOne(criteria);
    populations.forEach(function(pop) {
      query.populate(pop);
    });

Cheers! M

**Comment by eabovsky** [comment]
+1 - I also noticed this behavior. The instances always seem to auto-populate after a save() operation. This is obviously not a desired behavior as it's bad for performance and consistency...

**Comment by Verdier** [comment]
+1

**Comment by dmarcelino** [comment]
#954 may be a potential solution for this, I think. @nkibler7, what are your thoughts on this?

**Comment by nkibler7** [comment]
Looks exactly like what I was going for! I wasn't using the Sails.js PubSub feature, but my own implementation definitely requires optional association population. Any help or contribution to my pull request would be greatly appreciated!

**Comment by anasyb** [comment]
Model.save() shouldn't populate, this is an issue..

**Comment by devinivy** [comment]
As seen in #954, here are the relevant lines of code causing the population: https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L186-L190.  #954 is close, but requires a change in the function signature it proposed for `Model#save`.

**Comment by dmarcelino** [comment]
Regarding the default populate (or not) behaviour of `.save()` we should discuss changing it for `0.11`. Currently `.save()` populates by default and we can't change this in `0.10` as it may break applications. In the future I think `.save()` should not populate to be consistent with `.update()`, which does not populate by default.

**Comment by devinivy** [comment]
There are other options, too.  Firstly, (I believe) not every collection is populated– only those that were _already_ populated on the model instance.  For convenience, those populated records could be placed on the new model instance after save without requiring any extra queries.  I do understand why it makes sense in some ways to have the same populated records available before and after save.

**Comment by dmarcelino** [comment]
I see, that does some makes some sense. So...
* If the model instance has no populated associations: `.save()` should not populate any;
* If the model instance has some populated associations: `.save()` should populate already populated associations;
* If options are passed to `.save(options)`: the previous behaviours can be overwritten to whatever the user wants (PR #954 should take care of this).

I like this, this behaviour is consistent and probably is what users would expect. We need to write some tests to check what's happening now and to ensure it matches expectations.

**Comment by dmarcelino** [comment]
PR #1055 has been merged and improves current functionality. For `0.11` lets consider the proposed changes such as those the previous comment: https://github.com/balderdashy/waterline/issues/441#issuecomment-107227865.

**Comment by sailsbot** [comment]
Thanks for posting, @mphasize.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by dmarcelino** [comment]
As discussed previously:

> I see, that does make some sense. So...
> * If the model instance has no populated associations: `.save()` should not populate any;
> * If the model instance has some populated associations: `.save()` should populate already populated associations;
> * If options are passed to `.save(options)`: the previous behaviours can be overwritten to whatever the user wants (PR #1055 should take care of this).
> 
> I like this, this behaviour is consistent and probably is what users would expect. We need to write some tests to check what's happening now and to ensure it matches expectations.

So I'm reopening this as this will bring consistency to `.save()` behaviour.

**Comment by sailsbot** [comment]
Thanks for posting, @mphasize.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by didil** [comment]
.save() is still populating associations ? that's terrible and unexpected behavior , and I guess it can't be changed as this would be a breaking change ... ok I officially hate sails now

**Comment by maxoux** [comment]
Still populating by default, why we can't disable auto-populate on save function ?

**Comment by particlebanana** [comment]
@maxoux as of Waterline `0.11.0` the `save` method doesn't do any lookups.
